Cantera import error in python when built with system yaml-cpp

### Cantera version
2.5.0a3 (674c37ce3cbfd)

### Operating System
Linux (possibly any)

### Python/MATLAB version
Python 3.7.3 (possibly any)

### Expected Behavior
Compiling the Cantera python package with system `yaml-cpp` library should work. 

### Actual Behavior
If the cantera python package is built with system `yaml-cpp` then the import in python fails because the cantera cpython library is not linked against `libyaml-cpp.so`.

```python
>>> import cantera
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/mnt/data/devel/cantera/build/python/cantera/__init__.py", line 4, in <module>
    from ._cantera import *
ImportError: /mnt/data/devel/cantera/build/python/cantera/_cantera.cpython-37m-x86_64-linux-gnu.so: undefined symbol: _ZTIN4YAML13BadConversionE
 ```
The cantera python package was built on arch linux with these library versions

* scons 3.0.4
* python 3.7.3
* cython 0.29.11
* yaml-cpp 0.6.2

and system `yaml-cpp` installed in default include and lib paths, e.g.
```
/usr/include/yaml-cpp/yaml.h
/usr/lib/libyaml-cpp.so.0.6.2
```

The compilation completes successfully without errors. Also the system `yaml-cpp` is found correctly by scons: 
```bash
$ scons build prefix=/usr python_cmd=/usr/bin/python python_package=y system_fmt=y system_yamlcpp=y system_eigen=y extra_inc_dirs='/usr/include/eigen3'
scons: Reading SConscript files ...
INFO: SCons is using the following Python interpreter: /usr/bin/python2
Configuration variables read from 'cantera.conf' and command line:
    prefix = '/usr'
    python_package = 'y'
    python_cmd = '/usr/bin/python'
    system_eigen = 'y'
    system_fmt = 'y'
    system_yamlcpp = 'y'
    extra_inc_dirs = '/usr/include/eigen3'

Checking for C++ header file cmath... yes
Checking for C++ header file fmt/ostream.h... yes
INFO: Using system installation of fmt library.
INFO: Found fmt version 5.3.0
Checking for C++ header file yaml-cpp/yaml.h... yes
INFO: Using system installation of yaml-cpp library.
Checking for C++ header file gtest/gtest.h... yes
Checking for C++ header file gmock/gmock.h... yes
INFO: Using system installation of Googletest
Checking for C++ header file Eigen/Dense... yes
INFO: Using system installation of Eigen.
INFO: Found Eigen version 3.3.7
Checking whether __GLIBCXX__ is declared... yes
Checking whether _LIBCPP_VERSION is declared... no
Checking whether __clang__ is declared... no
INFO: Found Boost version 1.69
Checking whether boost::core::demangle is declared... yes
Checking for CVodeCreate(CV_BDF, CV_NEWTON) in C++ library sundials_cvodes... no
Checking for double x; log(x) in C library None... no
INFO: Using private installation of Sundials version 3.1.
INFO: Using 'gfortran' to build the Fortran 90 interface
INFO: Using NumPy version 1.16.4.
INFO: Using Cython version 0.29.11.
INFO: Building the full Python package for Python 3.7
scons: done reading SConscript files.
```

If the cantera python package is compiled alternatively with `yaml-cpp` from the git submodule, then everything will work as expected.

If I add `yaml-cpp` manually to the `env['LIBS']` variable somewhere in the `SConstruct` file, then the python package also works fine if it was built with system `yaml-cpp`.

Hi! Thanks for your patience and for reporting this issue. I can't reproduce this specific error on macOS, I'll have to try later on Linux. That said, I get an error during the linking of the Cantera `.dylib` (during the build process). After `brew install yaml-cpp`, I get

```bash
$ brew list yaml-cpp
/usr/local/Cellar/yaml-cpp/0.6.2_1/include/yaml-cpp/ (37 files)
/usr/local/Cellar/yaml-cpp/0.6.2_1/lib/libyaml-cpp.0.6.2.dylib
/usr/local/Cellar/yaml-cpp/0.6.2_1/lib/cmake/ (4 files)
/usr/local/Cellar/yaml-cpp/0.6.2_1/lib/pkgconfig/yaml-cpp.pc
/usr/local/Cellar/yaml-cpp/0.6.2_1/lib/ (2 other files)

$ cat cantera.conf
python_package = 'full'
f90_interface = 'n'
system_yamlcpp = 'y'
boost_inc_dir = '/Users/bryan/miniconda3/envs/cantera-dev/include'

$ scons build
```

and the error message complains about undefined symbols:

<details><pre><code>Undefined symbols for architecture x86_64:
  "YAML::InvalidNode::~InvalidNode()", referenced from:
      Cantera::AnyMap::fromYamlString(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in AnyMap.os
      Cantera::AnyMap::fromYamlFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in AnyMap.os
      YAML::as_if<Cantera::AnyMap, void>::operator()() const in AnyMap.os
      YAML::convert<Cantera::AnyMap>::decode(YAML::Node const&, Cantera::AnyMap&) in AnyMap.os
      YAML::as_if<Cantera::AnyValue, void>::operator()() const in AnyMap.os
      YAML::convert<Cantera::AnyValue>::decode(YAML::Node const&, Cantera::AnyValue&) in AnyMap.os
      (anonymous namespace)::elementTypes(YAML::Node const&) in AnyMap.os
      ...
  "YAML::BadConversion::~BadConversion()", referenced from:
      YAML::TypedBadConversion<Cantera::AnyMap>::~TypedBadConversion() in AnyMap.os
      YAML::TypedBadConversion<Cantera::AnyMap>::~TypedBadConversion() in AnyMap.os
      YAML::TypedBadConversion<Cantera::AnyValue>::~TypedBadConversion() in AnyMap.os
      YAML::TypedBadConversion<Cantera::AnyValue>::~TypedBadConversion() in AnyMap.os
      YAML::TypedBadConversion<bool>::~TypedBadConversion() in AnyMap.os
      YAML::TypedBadConversion<bool>::~TypedBadConversion() in AnyMap.os
      YAML::TypedBadConversion<std::__1::vector<long, std::__1::allocator<long> > >::~TypedBadConversion() in AnyMap.os
      ...
  "YAML::Dump(YAML::Node const&)", referenced from:
      YAML::convert<Cantera::AnyMap>::decode(YAML::Node const&, Cantera::AnyMap&) in AnyMap.os
  "YAML::Load(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      Cantera::AnyMap::fromYamlString(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in AnyMap.os
  "YAML::detail::node_data::end()", referenced from:
      YAML::convert<Cantera::AnyMap>::decode(YAML::Node const&, Cantera::AnyMap&) in AnyMap.os
      YAML::convert<Cantera::AnyValue>::decode(YAML::Node const&, Cantera::AnyValue&) in AnyMap.os
      (anonymous namespace)::elementTypes(YAML::Node const&) in AnyMap.os
      YAML::convert<std::__1::vector<long, std::__1::allocator<long> > >::decode(YAML::Node const&, std::__1::vector<long, std::__1::allocator<long> >&) in AnyMap.os
      YAML::convert<std::__1::vector<double, std::__1::allocator<double> > >::decode(YAML::Node const&, std::__1::vector<double, std::__1::allocator<double> >&) in AnyMap.os
      YAML::convert<std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >::decode(YAML::Node const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >&) in AnyMap.os
      YAML::convert<std::__1::vector<bool, std::__1::allocator<bool> > >::decode(YAML::Node const&, std::__1::vector<bool, std::__1::allocator<bool> >&) in AnyMap.os
      ...
  "YAML::detail::node_data::begin()", referenced from:
      YAML::convert<Cantera::AnyMap>::decode(YAML::Node const&, Cantera::AnyMap&) in AnyMap.os
      YAML::convert<Cantera::AnyValue>::decode(YAML::Node const&, Cantera::AnyValue&) in AnyMap.os
      (anonymous namespace)::elementTypes(YAML::Node const&) in AnyMap.os
      YAML::convert<std::__1::vector<long, std::__1::allocator<long> > >::decode(YAML::Node const&, std::__1::vector<long, std::__1::allocator<long> >&) in AnyMap.os
      YAML::convert<std::__1::vector<double, std::__1::allocator<double> > >::decode(YAML::Node const&, std::__1::vector<double, std::__1::allocator<double> >&) in AnyMap.os
      YAML::convert<std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >::decode(YAML::Node const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >&) in AnyMap.os
      YAML::convert<std::__1::vector<bool, std::__1::allocator<bool> > >::decode(YAML::Node const&, std::__1::vector<bool, std::__1::allocator<bool> >&) in AnyMap.os
      ...
  "YAML::convert<bool>::decode(YAML::Node const&, bool&)", referenced from:
      YAML::as_if<bool, void>::operator()() const in AnyMap.os
  "YAML::LoadFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      Cantera::AnyMap::fromYamlFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in AnyMap.os
  "typeinfo for YAML::InvalidNode", referenced from:
      Cantera::AnyMap::fromYamlString(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in AnyMap.os
      Cantera::AnyMap::fromYamlFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in AnyMap.os
      YAML::as_if<Cantera::AnyMap, void>::operator()() const in AnyMap.os
      YAML::convert<Cantera::AnyMap>::decode(YAML::Node const&, Cantera::AnyMap&) in AnyMap.os
      YAML::as_if<Cantera::AnyValue, void>::operator()() const in AnyMap.os
      YAML::convert<Cantera::AnyValue>::decode(YAML::Node const&, Cantera::AnyValue&) in AnyMap.os
      (anonymous namespace)::elementTypes(YAML::Node const&) in AnyMap.os
      ...
  "typeinfo for YAML::BadConversion", referenced from:
      typeinfo for YAML::TypedBadConversion<Cantera::AnyMap> in AnyMap.os
      typeinfo for YAML::TypedBadConversion<Cantera::AnyValue> in AnyMap.os
      typeinfo for YAML::TypedBadConversion<bool> in AnyMap.os
      typeinfo for YAML::TypedBadConversion<std::__1::vector<long, std::__1::allocator<long> > > in AnyMap.os
      typeinfo for YAML::TypedBadConversion<long> in AnyMap.os
      typeinfo for YAML::TypedBadConversion<std::__1::vector<double, std::__1::allocator<double> > > in AnyMap.os
      typeinfo for YAML::TypedBadConversion<double> in AnyMap.os
      ...
  "typeinfo for YAML::Exception", referenced from:
      GCC_except_table107 in AnyMap.os
      GCC_except_table108 in AnyMap.os
  "vtable for YAML::InvalidNode", referenced from:
      Cantera::AnyMap::fromYamlString(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in AnyMap.os
      Cantera::AnyMap::fromYamlFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in AnyMap.os
      YAML::InvalidNode::InvalidNode() in AnyMap.os
      YAML::as_if<Cantera::AnyMap, void>::operator()() const in AnyMap.os
      YAML::convert<Cantera::AnyMap>::decode(YAML::Node const&, Cantera::AnyMap&) in AnyMap.os
      YAML::as_if<Cantera::AnyValue, void>::operator()() const in AnyMap.os
      YAML::convert<Cantera::AnyValue>::decode(YAML::Node const&, Cantera::AnyValue&) in AnyMap.os
      ...
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "vtable for YAML::RepresentationException", referenced from:
      Cantera::AnyMap::fromYamlString(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in AnyMap.os
      Cantera::AnyMap::fromYamlFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in AnyMap.os
      YAML::InvalidNode::InvalidNode() in AnyMap.os
      YAML::as_if<Cantera::AnyMap, void>::operator()() const in AnyMap.os
      YAML::convert<Cantera::AnyMap>::decode(YAML::Node const&, Cantera::AnyMap&) in AnyMap.os
      YAML::as_if<Cantera::AnyValue, void>::operator()() const in AnyMap.os
      YAML::convert<Cantera::AnyValue>::decode(YAML::Node const&, Cantera::AnyValue&) in AnyMap.os
      ...
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "vtable for YAML::Exception", referenced from:
      YAML::Exception::Exception(YAML::Mark const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in AnyMap.os
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [build/lib/libcantera_shared.2.5.0.dylib] Error 1
ranlib build/lib/libcantera.a
scons: building terminated because of errors.
</code></pre></details>

--- a/SConstruct
+++ b/SConstruct
@@ -1620,6 +1620,10 @@ else:
 linkLibs.extend(env['sundials_libs'])
 linkSharedLibs.extend(env['sundials_libs'])
 
+if env['system_yamlcpp']:
+    linkLibs.append('yaml-cpp')
+    linkSharedLibs.append('yaml-cpp')
+
 #  Add LAPACK and BLAS to the link line
 if env['blas_lapack_libs']:
     linkLibs.extend(env['blas_lapack_libs'])

--- a/src/SConscript
+++ b/src/SConscript
@@ -84,8 +84,6 @@ if (localenv['system_sundials'] == 'y'):
 if localenv['blas_lapack_libs']:
     localenv.Append(LIBS=localenv['blas_lapack_libs'],
                     LIBPATH=localenv['blas_lapack_dir'])
-if localenv['system_fmt']:
-    localenv.Append(LIBS='fmt')
 
 if localenv['system_yamlcpp']:
     localenv.Append(LIBS=['yaml-cpp'])

--- a/src/SConscript
+++ b/src/SConscript
@@ -87,6 +87,9 @@ if localenv['blas_lapack_libs']:
 if localenv['system_fmt']:
     localenv.Append(LIBS='fmt')
 
+if localenv['system_yamlcpp']:
+    localenv.Append(LIBS=['yaml-cpp'])
+
 # Build the Cantera shared library
 if localenv['layout'] != 'debian':
     if localenv['renamed_shared_libraries']:

